### DISCUSSION TITLE
XAML Islands on Tkinter app?

### DISCUSSION THREAD

**Discussion by ElSaico** [pywinrt/pywinrt#72] [General]
On today's episode of "this might be either vastly overkill or a clean solution for our platform-specific woes":

One pain point in particular for the work I'm doing on a Tkinter app (clean up its theming kludge and replace with a pure ttk solution) is its Transparent theme.
It works by setting the transparency color the same as background, which causes obvious trouble on Windows 10 as it has no means to change the title bar to a specific color...

...unless [you set `AppWindow.title_bar.extends_content_into_title_bar` and make a custom control](https://learn.microsoft.com/en-us/windows/apps/develop/title-bar#full-customization), that is.

The current implementation disables the window decoration entirely and draws a replacement title bar with Tk; however, it requires a lot of special handling as we need to implement stuff like dragging and caption buttons, then apply convoluted workarounds for things like minimizing to systray.
There's a way to insert WinRT UI controls into Win32 applications - which they call "XAML Islands" - but [the existing documentation](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/using-the-xaml-hosting-api) is heavily centered around C++ and Visual Studio; would it be viable to do on PyWinRT, or would it require so many hoops that it makes no significant gain?

**Comment by ElSaico** [comment]
UPDATE: the documentation points to the UWP version of the API despite referring to the Windows App SDK one...

But here's an official sample https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/Samples/Islands

EDIT: I asked people on the App SDK repository for extra help (microsoft/WindowsAppSDK#4588), as this seems far from trivial.

**Reply by dlech** [reply]
>  the documentation points to the UWP version of the API despite referring to the Windows App SDK one...

PyWinRT includes both UWP (Windows SDK) APIs and Windows App SDK APIs. The former are in the `winrt.windows` namespace and the latter are in the `winrt.microsoft` namespace. This is why the documentation refers to both.

**Reply by ElSaico** [reply]
I was referring to the Microsoft documentation itself; looking again, it wasn't clear.

**Comment by dlech** [comment]
> would it require so many hoops that it makes no significant gain?

IMHO, yes. Based on the feedback you got, it sounds like you would need to create some custom component in C++ and then run `PyWinRT.exe` on it to create a Python wrapper to be able to use it.

Also, looking at the example you linked, there would need to be some sort of integration with the Tk event loop, which may or may not be possible depending on how much of the Windows internals Tk exposes.

And if you are using the Windows App SDK, it would require users to install the Windows App runtime in order to be able to use your Python app. If your app is only shipped with an installer that includes this, that would be OK, But if you want users to be able to `pip install` your app, then there would be the extra step of having users install the Windows App runtime on the first run.

**Reply by dlech** [reply]
On the other hand, if it could be done only with Windows SDK API and Direct2D as suggested in https://github.com/microsoft/WindowsAppSDK/discussions/4588#discussioncomment-10103052, it might not be too bad.

**Reply by ElSaico** [reply]
The app does have an installer, thankfully, but I hope the Composition path works.

**Reply by ElSaico** [reply]
Looking at the WUC-based solution, [the official example](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/using-the-visual-layer-with-win32) requires three objects at least:

- a `Compositor` - trivial
- a `DispatcherQueueController` - created by [a Win32 API call that returns a WinRT object](https://learn.microsoft.com/en-us/windows/win32/api/dispatcherqueue/nf-dispatcherqueue-createdispatcherqueuecontroller) *(wait, what)*
- a `DesktopWindowTarget` - created by an interop API that doesn't seem to be documented anywhere and involves a lot of typecasting
```cpp
#include <windows.ui.composition.interop.h>

<snip>

namespace abi = ABI::Windows::UI::Composition::Desktop;

auto interop = m_compositor.as<abi::ICompositorDesktopInterop>();
DesktopWindowTarget target{ nullptr };
check_hresult(interop->CreateDesktopWindowTarget(window, false, reinterpret_cast<abi::IDesktopWindowTarget**>(put_abi(target))));
```
Doesn't seem feasible ATM, unfortunately.

**Reply by ElSaico** [reply]
A possible alternative would be Win2D, but [its relevant `CanvasSwapChain` type](https://microsoft.github.io/Win2D/WinUI3/html/WithoutControls.htm) currently suffers from [the same interop troubles](https://microsoft.github.io/Win2D/WinUI3/html/Interop.htm).

Integration with WASDK's `AppWindow` is on the way (microsoft/Win2D#924), which eliminates the need for that convoluted path, but it's a 1.6 feature at the very best.

**Reply by dlech** [reply]
FYI, I started working on adding interop modules for the required extra functions at https://github.com/pywinrt/pywinrt/compare/main...dlech:pywinrt:comp-interop.

Unfortunately, it's not actually drawing on the Window for some reason. The dispatch queue seems to be working though since I can get timer callbacks to work with it.

If you want to compare it to https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/using-the-visual-layer-with-win32, that would be helpful. Maybe I just missed a step somewhere or have a typo?

**Reply by ElSaico** [reply]
Nice!
So, comparing with the docs, I found two things:

- on both samples, the Compositor's root visual is having `relative_offset_adjustment` set, where it should be `relative_size_adjustment`
- the pure Win32 sample is missing the `visuals.insert_at_top(visual)` line at the end of `add_element()`

Hope this is it - the interop functions themselves look fine to me.

**Reply by dlech** [reply]
> * on both samples, the Compositor's root visual is having `relative_offset_adjustment` set, where it should be `relative_size_adjustment`

That was it! Thanks for having a look!




https://github.com/user-attachments/assets/c555c688-a5cc-42f6-aa5a-9eddd196fc62

(don't know why the video look pink - IRL, the window is gray)

**Reply by ElSaico** [reply]
UPDATE: tried out 2.2.0 as soon as I could, and it threw me an error.

```
2024-08-20 16:27:29.786 UTC - ERROR - 21904:16792:16792 EDMarketConnector.main:2126: EDMC Critical Error: [WinError -2147024891] Access is denied
Traceback (most recent call last):
  File "C:\Users\<snip>\Projects\EDMarketConnector\EDMarketConnector.py", line 2124, in main
    app = AppWindow(root)
          ^^^^^^^^^^^^^^^
  File "C:\Users\<snip>\Projects\EDMarketConnector\EDMarketConnector.py", line 441, in __init__
    theme.initialize(self.w)
  File "C:\Users\<snip>\Projects\EDMarketConnector\theme.py", line 165, in initialize
    self.compositor = Compositor()
                      ^^^^^^^^^^^^
PermissionError: [WinError -2147024891] Access is denied
```

Weirdest thing is, the sample code actually works, so it must be something on the app flow.

**Reply by dlech** [reply]
https://stackoverflow.com/a/65134944/1976323 says that you need to set up the dispatcher first.

**Reply by ElSaico** [reply]
The dispatcher was set up but not assigned a variable - that fixed it, so the GC was likely pulling the rug on me. 😅

A few adjustments later and now it runs. Excellent!
